File: 4회차/79_word_search/79_word_search.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#define MAX_M 100
#define MAX_N 100

bool backtrack(char board[MAX_M][MAX_N], bool visited[MAX_M][MAX_N], int m, int n, int row, int col, char* word, int idx) {
    
    // found
    if (word[idx] == '\0') { return true; }

    // boundary check
    if (row < 0 || row >= m || col < 0 || col >= n) { return false; }
    // invalid path
    if (visited[row][col] || board[row][col] != word[idx]) { return false; }

    // find valid path from board[row][col]
    visited[row][col] = true;
    bool res = (backtrack(board, visited, m, n, row + 1, col, word, idx + 1) || // down
                backtrack(board, visited, m, n, row - 1, col, word, idx + 1) || // up
                backtrack(board, visited, m, n, row, col + 1, word, idx + 1) || // right
                backtrack(board, visited, m, n, row, col - 1, word, idx + 1));  // left
    visited[row][col] = false;
    return res;
}

bool exist(char board[MAX_M][MAX_N], int m, int n, char* word) {
    bool visited[MAX_M][MAX_N] = {false};

    for (int row = 0; row < m; ++row) {
        for (int col = 0; col < n; ++col) {
            if (board[row][col] == word[0]) {
                if (backtrack(board, visited, m, n, row, col, word, 0)) {
                    return true;
                }
            }
        }
    }
    return false;
}

File: 4회차/79_word_search/79_word_search.py
class Solution:
    def backtrack(self, board, visited, row, col, word, idx):
        m, n = len(board), len(board[0])

        # found
        if idx == len(word):
            return True

        # boundary check
        if row < 0 or row >= m or col < 0 or col >= n:
            return False
        # invalid path
        if visited[row][col] or board[row][col] != word[idx]:
            return False

        # find valid path from board[row][col]
        visited[row][col] = True
        res = (self.backtrack(board, visited, row + 1, col, word, idx + 1) or  # down
               self.backtrack(board, visited, row - 1, col, word, idx + 1) or  # up
               self.backtrack(board, visited, row, col + 1, word, idx + 1) or  # right
               self.backtrack(board, visited, row, col - 1, word, idx + 1))    # left
        visited[row][col] = False
        return res

    def exist(self, board, word):
        m, n = len(board), len(board[0])
        visited = [[False] * n for _ in range(m)]

        for row in range(m):
            for col in range(n):
                if board[row][col] == word[0]:
                    if self.backtrack(board, visited, row, col, word, 0):
                        return True
        return False